src/storage/src/table/streaming_table/state_table.rs

```diff
@@ -700,9 +700,69 @@ impl<S: StateStore> StateTable<S> {
                 .into_stream(),
         )
     }
+
+    /// This function scans rows from the relational table with specific `pk_prefix`.
+    pub async fn iter_with_pk<'a>(
+        &'a self,
+        pk_prefix: &'a Row,
+        epoch: u64,
+    ) -> StorageResult<RowStreamWithPk<'a, S>> {
+        let prefix_serializer = self.pk_serializer.prefix(pk_prefix.size());
+        let encoded_prefix = serialize_pk(pk_prefix, &prefix_serializer);
+        let encoded_key_range = range_of_prefix(&encoded_prefix);
+
+        // We assume that all usages of iterating the state table only access a single vnode.
+        // If this assertion fails, then something must be wrong with the operator implementation or
+        // the distribution derivation from the optimizer.
+        let vnode = self
+            .try_compute_vnode_by_pk_prefix(pk_prefix)
+            .expect("the records with `pk_prefix` should reside in the same vnode")
+            .to_be_bytes();
+        let encoded_key_range_with_vnode = prefixed_range(encoded_key_range, &vnode);
+
+        // Mem table iterator.
+        let mem_table_iter = self.mem_table.iter(encoded_key_range_with_vnode.clone());
+
+        // Storage iterator.
+        let storage_iter = {
+            // Construct prefix hint for prefix bloom filter.
+            let pk_prefix_indices = &self.pk_indices[..pk_prefix.size()];
+            let prefix_hint = {
+                if self.dist_key_indices.is_empty() || self.dist_key_indices != pk_prefix_indices {
+                    None
+                } else {
+                    Some([&vnode, &encoded_prefix[..]].concat())
+                }
+            };
+
+            trace!(
+                table_id = ?self.keyspace.table_id(),
+                ?prefix_hint, ?encoded_key_range_with_vnode, ?pk_prefix,
+                dist_key_indices = ?self.dist_key_indices, ?pk_prefix_indices,
+                "storage_iter_with_prefix"
+            );
+
+            StorageIterInner::<S>::new(
+                &self.keyspace,
+                prefix_hint,
+                encoded_key_range_with_vnode,
+                self.get_read_option(epoch),
+                self.data_types.clone(),
+            )
+            .await?
+            .into_stream()
+        };
+
+        Ok(
+            StateTableRowIter::new(mem_table_iter, storage_iter, self.data_types.clone())
+                .into_stream_with_pk(),
```

Comment:
Why duplicate 60 LoC just for this 1-liner change?

Reply:
strong +1. Please avoid duplicate code as possible since this is a very critical code path.